lib/AST/DeclContext.cpp

```diff
@@ -277,8 +281,8 @@ DeclContext *DeclContext::getInnermostSkippedFunctionContext() {
 }
 
 DeclContext *DeclContext::getParentForLookup() const {
-  if (isa<ProtocolDecl>(this) || isa<ExtensionDecl>(this)) {
-    // If we are inside a protocol or an extension, skip directly
+  if (isa<ExtensionDecl>(this)) {
```

Comment:
How about still returning getModuleScopeContext() here if we have an invalidly-nested protocol? That would fix your issue with superclass requirements.